### PR TITLE
Refactor `BuiltinView` to dynamically generate SQL

### PR DESCRIPTION
**Pull request by friendlymatthew**
## Addresses: https://github.com/MaterializeInc/materialize/issues/17892

# This PR does the following: 
1. added an optional `column_defs` field to the `BuiltinView` struct to accommodate column definitions for views where explicit column listing is required
2. implement the `create_sql()` function in the `BuiltinView` struct
3. revised all instances of `BuiltinView` throughout the codebase, ensuring the `sql` field is updated to contain only the query part
4. dynamically generates the correct SQL upon opening the catalog

### CODE REVIEW

File: src/catalog/src/builtin.rs
Comment (friendlymatthew) [inline src/catalog/src/builtin.rs:2511]: Upon building, this specific view errs:

 ```bash
internal error: failed to load bootstrap view:
mz_objects
error:
PlanError(UnknownColumn { table: None, column: ColumnName("type"), similar: [] })
make sure that the schema name is specified in the builtin view's create sql statement.
```

**Comment by friendlymatthew** [comment]
Investigating why `legacy-upgrade-tests` fail. Plus, I printed out the SQL statements from every BuiltinView. [Here's the diff (new vs. original)](https://www.diffchecker.com/BXzzhRQ4/)

As suspected, the majority of diffs regard to white space. There are couple of statements that are misaligned from the schema and name.

**Comment by friendlymatthew** [comment]
FWIW, when I revert to the main branch, `bin/environmentd` outputs this error:

Screen Shot 2023-11-12 at 2 10 35 PM

File: src/catalog/src/builtin.rs
Comment (friendlymatthew) [inline src/catalog/src/builtin.rs:2529]: Schema differs here. (`mz_catalog` vs. `MZ_INTERNAL_SCHEMA` or `mz_internal`)

File: src/catalog/src/builtin.rs
Comment (friendlymatthew) [inline src/catalog/src/builtin.rs:3803]: The name differs here. `mz_dataflow_operator_parents_per_worker` vs. `mz_operator_parents_per_worker`.

File: src/catalog/src/builtin.rs
Comment (friendlymatthew) [inline src/catalog/src/builtin.rs:3829]: the name differs here. `mz_dataflow_operator_parents` vs. `mz_operator_parents`

File: src/catalog/src/builtin.rs
Comment (friendlymatthew) [inline src/catalog/src/builtin.rs:2902]: This relates to why the tests fail: 

The sql below doesn't properly transform:

expected: 
`mz_internal.mz_type_pg_metadata.typinput::pg_catalog.regproc AS typinput,`

but instead:
`(CASE mztype WHEN 'a' THEN 'array_in' ELSE NULL END)::pg_catalog.regproc AS typinput,`

I'm a bit confused to why this happens, as you can see, I don't touch anything other than skimming the top off.

Comment (friendlymatthew) [reply src/catalog/src/builtin.rs:2902]: [LINE 935](https://www.diffchecker.com/BXzzhRQ4/)

**Comment by jkosh44** [comment]
> There are couple of statements that are misaligned from the schema and name.

Aha, that's the exact type of thing I was hoping to uncover/prevent in the future with this task. Before merging this, there are some implications that we have to think through of how these objects currently look in prod and how this change will affect them. I.e., we need to make sure they migrate cleanly. You're welcome to try and figure this out, but it's definitely beyond the scope of a good first issue. I'll take a look on Monday.

> FWIW, when I revert to the main branch, `bin/environmentd` outputs the same error as `legacy-upgrade-tests`:
> Screen Shot 2023-11-12 at 2 10 35 PM

You probably want to run `bin/environmentd --reset` otherwise it will use the data created from a previous run.

**Comment by jkosh44** [comment]
> Before merging this, there are some implications that we have to think through of how these objects currently look in prod and how this change will affect them. I.e., we need to make sure they migrate cleanly. You're welcome to try and figure this out, but it's definitely beyond the scope of a good first issue. I'll take a look on Monday.

I couldn't help myself. Here's the full list of updated views according to your diff:

- `MZ_OBJECT_FULLY_QUALIFIED_NAMES`: schema is `mz_internal` but CREATE SQL is `mz_catalog`.
- `MZ_DATAFLOW_OPERATOR_PARENTS_PER_WORKER`: name is `mz_dataflow_operator_parents_per_worker` but CREATE SQL is `mz_operator_parents_per_worker`.
- `MZ_DATAFLOW_OPERATOR_PARENTS`: Name is `mz_dataflow_operator_parents` but the CREATE SQL is `mz_operator_parents`.


For `MZ_OBJECT_FULLY_QUALIFIED_NAMES`, it turns out that the schema in the CREATE SQL is ignored, so this is in fact in `mz_internal`, so there's no migration needed. For the other two, they are in `mz_internal`, and we are ok to remove/change any object in `mz_internal`. So all the differences are fine.

File: src/catalog/src/builtin.rs
Comment (jkosh44) [reply src/catalog/src/builtin.rs:2902]: Hmm, weird. Line 935 in that diff is for `pg_type`, but the error and this view is `pg_depend`. I'm not sure I understand the connection yet and why there is a difference. It's also interesting that this only happens in the legacy upgrade tests. It implies the problem only happens after/during an upgrade.

File: src/catalog/src/builtin.rs
Comment (friendlymatthew) [reply src/catalog/src/builtin.rs:4078]: Should be fixed with my next commit!

**Comment by friendlymatthew** [comment]
> So all the differences are fine.

Sounds good. Will leave it as is then

File: src/catalog/src/builtin.rs
Comment (jkosh44) [inline src/catalog/src/builtin.rs:165]: We'll have to update the `impl Fingerprint` to use this method. It's down on line 274

```Rust
impl Fingerprint for &BuiltinView {
    fn fingerprint(&self) -> String {
        self.sql.to_string()
    }
}
```

instead of `self.sql` it should be `self.create_sql()`.

Comment (jkosh44) [reply src/catalog/src/builtin.rs:165]: It's possible that it might also help with the migration error that we're seeing.

Comment (friendlymatthew) [reply src/catalog/src/builtin.rs:165]: Interesting! I should've looked more closely with the `Fingerprint` trait. Thank you

Comment (jkosh44) [reply src/catalog/src/builtin.rs:165]: I tried it out locally and it doesn't seem to help. The panic is happening in the builtin migration framework, which is responsible for migrating system views (and other objects) when they're changed between releases. We use the `Fingerprint` trait to determine when an object has been changed. As you can see from the trait implementation, even if the white-space is changed in the CREATE SQL statement for a view, it will be considered changed. This is probably the largest migration we've had in awhile and  I think something in the framework itself isn't working properly.

**Comment by friendlymatthew** [comment]
This doesn't resolve the issue, but I figured I could do my best with the formatting.

My most recent commit tries to minimize the diffs in SQL statements.

[Here is the new diff](https://www.diffchecker.com/hvI1GIq0/), as you can see it resolves about half of the clashing diffs.

**Comment by friendlymatthew** [comment]
This doesn't resolve the legacy upgrade issue, but I wanted to see how many diffs of SQL lines I could minimize.

By adding an extra attribute in `BuiltinView` and writing extra logic in `create_sql()`, we get down to only 6 diffs. 
### [See Diff Checker](https://www.diffchecker.com/REfEbLzw/)

Screen Shot 2023-11-13 at 1 54 00 AM

<br />
<br />

*But if this is not needed or might sacrifice readability or clearness, I'll be happy to roll back to [the previous commit](https://github.com/MaterializeInc/materialize/pull/23125/commits/9a16655f545f9b09b8fe3482a27abfd0805de303)*

**Comment by jkosh44** [comment]
> This doesn't resolve the legacy upgrade issue, but I wanted to see how many diffs of SQL lines I could minimize.
> 
> By adding an extra attribute in `BuiltinView` and writing extra logic in `create_sql()`, we get down to only 6 diffs.
> ### [See Diff Checker](https://www.diffchecker.com/REfEbLzw/)
> 
> _But if this is not needed or might sacrifice readability or clearness, I'll be happy to roll back to [the previous commit](https://github.com/MaterializeInc/materialize/pull/23125/commits/9a16655f545f9b09b8fe3482a27abfd0805de303)_

It looks like your white-space change resolved the panic. I did some additional investigating, and it's definitely a builtin migration bug. This PR should fail with the exact same panic: https://github.com/MaterializeInc/materialize/pull/23151, at least it does locally. I'm very hesitant to add in all this logic around manipulating white-space just to try and hide a bug somewhere else. Let me try and resolve this bug and then if I get it fixed you can rebase over the fix.

**Comment by friendlymatthew** [comment]
> It looks like your white-space change resolved the panic. I did some additional investigating, and it's definitely a builtin migration bug. This PR should fail with the exact same panic: #23151, at least it does locally. I'm very hesitant to add in all this logic around manipulating white-space just to try and hide a bug somewhere else. Let me try and resolve this bug and then if I get it fixed you can rebase over the fix.

I agree -- definitely not the ideal solution. Very curious about the bug, so will keep my eye on your PR.

**Comment by jkosh44** [comment]
@friendlymatthew here's a bit of an update. I've put together a more robust solution here: https://github.com/MaterializeInc/materialize/pull/23168. Locally it seems to work, but we'll see what happens in CI. I'm not really in love with the PR for the following reasons:

- There seems to be a lot of code duplication across the references fields/methods and uses fields/methods.
- I had to add a lot of cloning, which could probably be avoided.
- The test I added is a bit scary. I'm hoping there's some `cfg` Rust magic I can use to avoid the mutex outside of tests.

Unfortunately, I have to put it down for a bit and focus on other things. If you're feeling really ambitious, you can try to clean it up, but it's definitely far beyond the scope of a first time contributor.

In the meantime I've opened the following two follow-up issues:

- https://github.com/MaterializeInc/materialize/issues/23200
- https://github.com/MaterializeInc/materialize/issues/23201

These both should not be blocked by https://github.com/MaterializeInc/materialize/issues/23153. The indexes shouldn't trigger a builtin migration large enough to cause issues. If you're looking for something to do, these are both good issue for external contributors.

**Comment by friendlymatthew** [comment]
Hey @jkosh44, I understand. Happy to work on the other two issues. Fwiw, I have a [PR up](https://github.com/MaterializeInc/materialize/pull/23128) that addresses [this issue](https://github.com/MaterializeInc/materialize/issues/20166). But I'm currently facing a blocker with my `cargo test`, so if the approach is incorrect, I can gladly close that issue.

**Comment by jkosh44** [comment]
Hi @friendlymatthew the bug that was blocking this PR is now fixed! We can now work towards merging this PR, which would require the following steps:

1. Revert `5e2bf34ff2c33d0985ac770182e922a233cf5e50`.
2. Revert `9a16655f545f9b09b8fe3482a27abfd0805de303`.
3. Pull/Rebase `main`.
4. Resolve merge conflicts.

I'd be happy to do that for you, but I don't want to deprive you of the opportunity if you were potentially excited to practice some git.

**Comment by friendlymatthew** [comment]
> I'd be happy to do that for you, but I don't want to deprive you of the opportunity if you were potentially excited to practice some git.

Awesome! And yes, I'd definitely prefer it if I did this :). Definitely curious to see the bug fix so will read that as well.

P.S. if there is any more grunt work, please defer to me. I'll also look for other issues to work on

File: src/catalog/src/builtin.rs
Comment (jkosh44) [inline src/catalog/src/builtin.rs:3645]: Looks like you accidentally deleted these changes as part of resolving the merge conflicts.

**Comment by jkosh44** [comment]
> Awesome! And yes, I'd definitely prefer it if I did this :). Definitely curious to see the bug fix so will read that as well.
> 
> P.S. if there is any more grunt work, please defer to me. I'll also look for other issues to work on

Sounds good!

**Review by jkosh44 — APPROVED** [review]
LGTM!

**Comment by friendlymatthew** [comment]
> LGTM!

Sorry, I had to remove an unused import

**Pull request merged**
